# Explore health dataset

## Import libraries

In [1]:
import pandas as pd 

## Loading datasets

In [2]:
dataset1 = pd.read_excel('/Users/vishalsaxena/Documents/DS_Workspace/healthdax/data/raw/Health Dataset 1.xlsm')
dataset1

Patient_Number  Blood_Pressure_Abnormality  Level_of_Hemoglobin  \
0                  1                           1                11.28   
1                  2                           0                 9.75   
2                  3                           1                10.79   
3                  4                           0                11.00   
4                  5                           1                14.17   
...              ...                         ...                  ...   
1995            1996                           1                10.14   
1996            1997                           1                11.77   
1997            1998                           1                16.91   
1998            1999                           0                11.15   
1999            2000                           1                11.36   

      Genetic_Pedigree_Coefficient  Age  BMI  Sex  Pregnancy  Smoking  \
0                             0.90   34   23    1        1.0        0   
1                             0.23   54   33    1        NaN        0   
2                             0.91   70   49    0        NaN        0   
3                             0.43   71   50    0        NaN        0   
4                             0.83   52   19    0        NaN        0   
...                            ...  ...  ...  ...        ...      ...   
1995                          0.02   69   26    1        NaN        1   
1996                          1.00   24   45    1        1.0        1   
1997                          0.22   18   42    0        NaN        0   
1998                          0.72   46   45    1        NaN        1   
1999                          0.09   41   45    0        NaN        0   

      salt_content_in_the_diet  alcohol_consumption_per_day  Level_of_Stress  \
0                        48071                          NaN                2   
1                        25333                        205.0                3   
2                        29465                         67.0                2   
3                         7439                        242.0                1   
4                        49644                        397.0                2   
...                        ...                          ...              ...   
1995                     47568                        144.0                3   
1996                      8063                          NaN                3   
1997                     24753                          NaN                2   
1998                     15275                        253.0                3   
1999                     30463                        230.0                1   

      Chronic_kidney_disease  Adrenal_and_thyroid_disorders  
0                          1                              1  
1                          0                              0  
2                          1                              0  
3                          0                              0  
4                          0                              0  
...                      ...                            ...  
1995                       1                              0  
1996                       1                              1  
1997                       1                              1  
1998                       0                              0  
1999                       1                              0  

[2000 rows x 14 columns]

In [3]:
dataset2 = pd.read_excel('/Users/vishalsaxena/Documents/DS_Workspace/healthdax/data/raw/Health Dataset 2.xlsm')
dataset2

Patient_Number  Day_Number  Physical_activity
0                   1           1            23590.0
1                   1           2            10411.0
2                   1           3             7815.0
3                   1           4            12366.0
4                   1           5                NaN
...               ...         ...                ...
19995            2000           6            34220.0
19996            2000           7            36223.0
19997            2000           8            28034.0
19998            2000           9                NaN
19999            2000          10            22080.0

[20000 rows x 3 columns]

## Datasets analysis

In [4]:
dataset1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 14 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Patient_Number                 2000 non-null   int64  
 1   Blood_Pressure_Abnormality     2000 non-null   int64  
 2   Level_of_Hemoglobin            2000 non-null   float64
 3   Genetic_Pedigree_Coefficient   1908 non-null   float64
 4   Age                            2000 non-null   int64  
 5   BMI                            2000 non-null   int64  
 6   Sex                            2000 non-null   int64  
 7   Pregnancy                      442 non-null    float64
 8   Smoking                        2000 non-null   int64  
 9   salt_content_in_the_diet       2000 non-null   int64  
 10  alcohol_consumption_per_day    1758 non-null   float64
 11  Level_of_Stress                2000 non-null   int64  
 12  Chronic_kidney_disease         2000 non-null   i

### Initial analysis of dataset1
- The dataset consist of 2000 rows and 14 columns
- The type of all variables in the dataset are in numerical format i.e integer/float

In [5]:
dataset2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 3 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Patient_Number     20000 non-null  int64  
 1   Day_Number         20000 non-null  int64  
 2   Physical_activity  16159 non-null  float64
dtypes: float64(1), int64(2)
memory usage: 468.9 KB


### Initial analysis of dataset2
- The dataset consist of 20000 rows and 3 columns
- The type of all variables in the dataset are numerical format

## EDA of dataset1

### Examining missing values

In [6]:
dataset1.isna().sum().to_frame()

0
Patient_Number                    0
Blood_Pressure_Abnormality        0
Level_of_Hemoglobin               0
Genetic_Pedigree_Coefficient     92
Age                               0
BMI                               0
Sex                               0
Pregnancy                      1558
Smoking                           0
salt_content_in_the_diet          0
alcohol_consumption_per_day     242
Level_of_Stress                   0
Chronic_kidney_disease            0
Adrenal_and_thyroid_disorders     0

#### Missing values analysis
- Based on missing values dataframe of dataset1
    - 1558 missing values in Pregnancy
        - handling considering 1 is female and male is 0, then fill pregancy with 0 for male 
        - incase female fill with -1 (unknown)
    - 242 missing values in alcohol_consumption_per_day
        - filling with 0 considering non-drinker
    - 92 missing values in Genetic_Pedigree_Coefficients
        - filling with median considering numeric variable
- Based on missing values dataframe of dataset2
    - There are missing values in Physical_activity
    - 3841 missing values in Physical_actvity

#### Handle missing values (pregnancy variable)

In [7]:
dataset1.loc[(dataset1['Sex'] == 0) & (dataset1['Pregnancy'].isnull()), 'Pregnancy'] = 0
dataset1.loc[(dataset1['Sex'] == 1) & (dataset1['Pregnancy'].isnull()), 'Pregnancy'] = -1
dataset1.isna().sum().to_frame()

0
Patient_Number                   0
Blood_Pressure_Abnormality       0
Level_of_Hemoglobin              0
Genetic_Pedigree_Coefficient    92
Age                              0
BMI                              0
Sex                              0
Pregnancy                        0
Smoking                          0
salt_content_in_the_diet         0
alcohol_consumption_per_day    242
Level_of_Stress                  0
Chronic_kidney_disease           0
Adrenal_and_thyroid_disorders    0

#### Handle missing values (alcohol_consumption_per_day)

In [8]:
dataset1['alcohol_consumption_per_day'] = dataset1['alcohol_consumption_per_day'].fillna(0)
dataset1.isna().sum().to_frame()

0
Patient_Number                  0
Blood_Pressure_Abnormality      0
Level_of_Hemoglobin             0
Genetic_Pedigree_Coefficient   92
Age                             0
BMI                             0
Sex                             0
Pregnancy                       0
Smoking                         0
salt_content_in_the_diet        0
alcohol_consumption_per_day     0
Level_of_Stress                 0
Chronic_kidney_disease          0
Adrenal_and_thyroid_disorders   0

#### Handle missing values (Genetic_Pedigree_Coefficient)

In [9]:
median_GPC = dataset1['Genetic_Pedigree_Coefficient'].median()
dataset1['Genetic_Pedigree_Coefficient'] = dataset1['Genetic_Pedigree_Coefficient'].fillna(median_GPC)
dataset1.isna().sum().to_frame()

0
Patient_Number                 0
Blood_Pressure_Abnormality     0
Level_of_Hemoglobin            0
Genetic_Pedigree_Coefficient   0
Age                            0
BMI                            0
Sex                            0
Pregnancy                      0
Smoking                        0
salt_content_in_the_diet       0
alcohol_consumption_per_day    0
Level_of_Stress                0
Chronic_kidney_disease         0
Adrenal_and_thyroid_disorders  0

In [10]:
dataset2.isna().sum().to_frame()

0
Patient_Number        0
Day_Number            0
Physical_activity  3841

### Examining unique values

In [11]:
# dataset1['Pregnancy'].value_counts()
unique_count = []
for col_name in dataset1.columns:
    unq_val_cnt = dataset1[col_name].value_counts().count()
    unique_count.append(unq_val_cnt)
# unique_count
pd.DataFrame(unique_count, index=dataset1.columns, columns=['Unique Count'])

Unique Count
Patient_Number                         2000
Blood_Pressure_Abnormality                2
Level_of_Hemoglobin                     757
Genetic_Pedigree_Coefficient            101
Age                                      58
BMI                                      41
Sex                                       2
Pregnancy                                 3
Smoking                                   2
salt_content_in_the_diet               1945
alcohol_consumption_per_day             488
Level_of_Stress                           3
Chronic_kidney_disease                    2
Adrenal_and_thyroid_disorders             2

#### Unique Values Analysis 
- Based on Unique Count Dataframe, we consider
    - **Variables with few unique values as categorical and high unique values as numeric variables**
    - hence the Numeric variables are 
        - Level_of_Hemoglobin 
        - Genetic_Pedigree_Coefficient
        - Age
        - BMI 
        - salt_content_in_the_diet
        - alcohol_consumption_per_day
    - and Categorical variables are 
        - Sex
        - Pregnancy
        - Smoking
        - Level_of_Stress
        - Chronic_kidney_disease
        - Adrenal_and_thyroid_disorders

### Separating numeric and categorical variables

In [12]:
numeric_var = ["Level_of_Hemoglobin", "Genetic_Pedigree_Coefficient", "Age", "BMI", "salt_content_in_the_diet",
              "alcohol_consumption_per_day"]
categoric_var = ["Sex", "Pregnancy", "Smoking", "Level_of_Stress", "Chronic_kidney_disease", 
                 "Adrenal_and_thyroid_disorders"]

### Statistical analysis numeric variables

In [13]:
dataset1[numeric_var].describe()

Level_of_Hemoglobin  Genetic_Pedigree_Coefficient          Age  \
count          2000.000000                   2000.000000  2000.000000   
mean             11.710035                      0.494595    46.558500   
std               2.186701                      0.284945    17.107832   
min               8.100000                      0.000000    18.000000   
25%              10.147500                      0.250000    32.000000   
50%              11.330000                      0.490000    46.000000   
75%              12.945000                      0.730000    62.000000   
max              17.560000                      1.000000    75.000000   

               BMI  salt_content_in_the_diet  alcohol_consumption_per_day  
count  2000.000000               2000.000000                  2000.000000  
mean     30.081500              24926.097000                   220.636500  
std      11.761208              14211.692586                   157.614106  
min      10.000000                 22.000000                     0.000000  
25%      20.000000              13151.750000                    79.000000  
50%      30.000000              25046.500000                   212.500000  
75%      40.000000              36839.750000                   360.250000  
max      50.000000              49976.000000                   499.000000

In [15]:
dataset1[numeric_var].skew().to_frame()

0
Level_of_Hemoglobin           0.657066
Genetic_Pedigree_Coefficient  0.017873
Age                           0.021178
BMI                          -0.017555
salt_content_in_the_diet     -0.021798
alcohol_consumption_per_day   0.090886

## EDA of dataset2

### Examining missing values

In [22]:
dataset2.head(40)

Patient_Number  Day_Number  Physical_activity
0                1           1            23590.0
1                1           2            10411.0
2                1           3             7815.0
3                1           4            12366.0
4                1           5                NaN
5                1           6            30319.0
6                1           7            46225.0
7                1           8            40988.0
8                1           9            10540.0
9                1          10            27409.0
10               2           1            34928.0
11               2           2                NaN
12               2           3            39409.0
13               2           4            19205.0
14               2           5            26894.0
15               2           6            11518.0
16               2           7                NaN
17               2           8            19584.0
18               2           9            37076.0
19               2          10            35095.0
20               3           1            31479.0
21               3           2             8537.0
22               3           3                NaN
23               3           4            20167.0
24               3           5            26392.0
25               3           6                NaN
26               3           7            32602.0
27               3           8                NaN
28               3           9            28268.0
29               3          10            12635.0
30               4           1            31666.0
31               4           2                NaN
32               4           3                NaN
33               4           4            39404.0
34               4           5            24287.0
35               4           6            30045.0
36               4           7            23975.0
37               4           8            27817.0
38               4           9                NaN
39               4          10            25242.0

In [18]:
dataset2.isna().sum().to_frame()

0
Patient_Number        0
Day_Number            0
Physical_activity  3841

In [30]:
dataset2.groupby('Patient_Number')['Day_Number'].count() 

Patient_Number
1       10
2       10
3       10
4       10
5       10
        ..
1996    10
1997    10
1998    10
1999    10
2000    10
Name: Day_Number, Length: 2000, dtype: int64

In [38]:
dataset2['is_null'] = dataset2['Physical_activity'].isna()

result = dataset2.groupby('Patient_Number')['is_null'] \
  .sum() \
  .reset_index(name='null_count')
result

Patient_Number  null_count
0                  1           1
1                  2           2
2                  3           3
3                  4           3
4                  5           4
...              ...         ...
1995            1996           2
1996            1997           1
1997            1998           1
1998            1999           3
1999            2000           1

[2000 rows x 2 columns]

In [39]:
result['missing_percent'] = (
    result['null_count'] / 10
) * 100
result

Patient_Number  null_count  missing_percent
0                  1           1             10.0
1                  2           2             20.0
2                  3           3             30.0
3                  4           3             30.0
4                  5           4             40.0
...              ...         ...              ...
1995            1996           2             20.0
1996            1997           1             10.0
1997            1998           1             10.0
1998            1999           3             30.0
1999            2000           1             10.0

[2000 rows x 3 columns]

In [40]:
high_missing = result[result['null_count'] >= 3]
high_missing

Patient_Number  null_count  missing_percent
2                  3           3             30.0
3                  4           3             30.0
4                  5           4             40.0
8                  9           4             40.0
9                 10           4             40.0
...              ...         ...              ...
1977            1978           4             40.0
1979            1980           3             30.0
1981            1982           5             50.0
1988            1989           3             30.0
1998            1999           3             30.0

[598 rows x 3 columns]

In [42]:
patient_activity_df = dataset2.groupby('Patient_Number')['Physical_activity'].mean().reset_index()
patient_activity_df

Patient_Number  Physical_activity
0                  1       23295.888889
1                  2       27963.625000
2                  3       22868.571429
3                  4       28919.428571
4                  5       21660.333333
...              ...                ...
1995            1996       29824.250000
1996            1997       32334.111111
1997            1998       20377.000000
1998            1999       26548.428571
1999            2000       26843.111111

[2000 rows x 2 columns]

In [44]:
merge_health_df = pd.merge(dataset1, patient_activity_df, on='Patient_Number')
merge_health_df

Patient_Number  Blood_Pressure_Abnormality  Level_of_Hemoglobin  \
0                  1                           1                11.28   
1                  2                           0                 9.75   
2                  3                           1                10.79   
3                  4                           0                11.00   
4                  5                           1                14.17   
...              ...                         ...                  ...   
1995            1996                           1                10.14   
1996            1997                           1                11.77   
1997            1998                           1                16.91   
1998            1999                           0                11.15   
1999            2000                           1                11.36   

      Genetic_Pedigree_Coefficient  Age  BMI  Sex  Pregnancy  Smoking  \
0                             0.90   34   23    1        1.0        0   
1                             0.23   54   33    1       -1.0        0   
2                             0.91   70   49    0        0.0        0   
3                             0.43   71   50    0        0.0        0   
4                             0.83   52   19    0        0.0        0   
...                            ...  ...  ...  ...        ...      ...   
1995                          0.02   69   26    1       -1.0        1   
1996                          1.00   24   45    1        1.0        1   
1997                          0.22   18   42    0        0.0        0   
1998                          0.72   46   45    1       -1.0        1   
1999                          0.09   41   45    0        0.0        0   

      salt_content_in_the_diet  alcohol_consumption_per_day  Level_of_Stress  \
0                        48071                          0.0                2   
1                        25333                        205.0                3   
2                        29465                         67.0                2   
3                         7439                        242.0                1   
4                        49644                        397.0                2   
...                        ...                          ...              ...   
1995                     47568                        144.0                3   
1996                      8063                          0.0                3   
1997                     24753                          0.0                2   
1998                     15275                        253.0                3   
1999                     30463                        230.0                1   

      Chronic_kidney_disease  Adrenal_and_thyroid_disorders  Physical_activity  
0                          1                              1       23295.888889  
1                          0                              0       27963.625000  
2                          1                              0       22868.571429  
3                          0                              0       28919.428571  
4                          0                              0       21660.333333  
...                      ...                            ...                ...  
1995                       1                              0       29824.250000  
1996                       1                              1       32334.111111  
1997                       1                              1       20377.000000  
1998                       0                              0       26548.428571  
1999                       1                              0       26843.111111  

[2000 rows x 15 columns]

In [46]:
merge_health_df.to_csv('/Users/vishalsaxena/Documents/DS_Workspace/healthdax/data/processed/patient_health_data.csv')